app/models/tag.rb

```diff
@@ -23,6 +23,80 @@ class Tag < ActiveRecord::Base
   # the order is important, and it is the order in which they appear in the tag wrangling interface
   USER_DEFINED = ['Fandom', 'Character', 'Relationship', 'Freeform']
 
+  def self.write_redis_to_database
+    REDIS_GENERAL.smembers("tag_update").each_slice(1000) do |batch|
+      Tag.transaction do
+        batch.each do |id|
+          value, large_tag = REDIS_GENERAL.mget("tag_update_#{id}_value", "tag_update_#{id}_large_tag")
+          sql = []
+          sql.push("taggings_count_cache = #{value}") unless value.blank?
+          sql.push("large_tag = #{large_tag}") unless large_tag.blank?
+          Tag.where(id: id).update_all(sql.join(",")) unless sql.empty?
+        end
+        REDIS_GENERAL.srem("tag_update", batch)
+      end
+    end
+  end
+
+  def check_if_large_tag(time_in_cache)
+    # A large tag is one which is periodically refreshed by a background task.
+    # The task is defined in config/resque_schedule.yml
+    # This refreshes the large tags every 30 minutes. It seemed sensible to
+    # only allow tags which stay in the cache longer than the periodic refresh
+    # period.
+    target_state = (time_in_cache >= 40)
+    return if large_tag == target_state
+    write_taggings_to_redis(nil, target_state)
+  end
+
+  def taggings_count_expiry(count)
+    # What we are trying to do here is work out a resonable amount of time for a work to be cached for
+    # This should take the number of taggings and divide it by TAGGINGS_COUNT_CACHE_DIVISOR  ( defaults to 2000 )
+    # such that for example 2000, would be naturally be tagged for one minute while 140,000 would be cached for
+    # 70 minutes. However we then apply a filter such that the minimum amount of time we will cache something for
+    # would be TAGGINGS_COUNT_MIN_TIME ( defaults to 3 minutes ) and the maximum amount of time would be
+    # TAGGINGS_COUNT_MAX_TIME ( defaulting to an hour ).
+    expiry_time = count / (ArchiveConfig.TAGGINGS_COUNT_CACHE_DIVISOR || 2000)
+    [[expiry_time, (ArchiveConfig.TAGGINGS_COUNT_MIN_TIME || 3)].max, (ArchiveConfig.TAGGINGS_COUNT_MAX_TIME || 60)].min
+  end
+
+  def taggings_count_cache_key
+    "/v1/taggings_count/#{id}"
+  end
+
+  def write_taggings_to_redis(value, large_tag)
+    REDIS_GENERAL.sadd("tag_update", id)
+    REDIS_GENERAL.mset("tag_update_#{id}_value", value, "tag_update_#{id}_large_tag", large_tag)
+    value
+  end
+
+  def taggings_count=(value)
+    expiry_time = taggings_count_expiry(value)
+    # Only write to the cache if there are more than TAGGINGS_COUNT_MIN_CACHE_COUNT ( defaults to 1,000 ) uses.
+    Rails.cache.write(taggings_count_cache_key, value, race_condition_ttl: 10, expires_in: expiry_time.minutes) if value >= (ArchiveConfig.TAGGINGS_COUNT_MIN_CACHE_COUNT || 1000)
+    check_if_large_tag(expiry_time)
+    write_taggings_to_redis(value, nil)
```

Comment:
I believe this will overwrite the value stored in REDIS for the key `"tag_update_#{id}_large_tag"` (which may have just been set by the previous line's call to `check_if_large_tag`). From my testing, `redis-rb` treats `nil` as being equivalent to the empty string, with no special handling. It makes sense, since all REDIS keys & values are strings, and `nil.to_s` is the empty string.

Since `write_taggings_to_redis` is never called with both `value` and `large_tag` as non-nil arguments, I'd suggest separating `write_taggings_to_redis` into two functions: one for storing the taggings count, and one for storing the value of large_tag.

Alternatively, it might be more efficient to avoid storing the value of `large_tag` on REDIS, because it can be computed directly from the `taggings_count`. You could consider modifying `Tag.write_redis_to_database` to compute the `large_tag` value from the new `taggings_count` value, and write both to MySQL at the same time. (You might even be able eliminate the `large_tag` column altogether and just search for tags with a large value of `taggings_count_cache` -- there's no index either way, so I believe MySQL would have to examine all tags regardless.)